tests/add_test.py

```diff
@@ -0,0 +1,228 @@
+            "-fullwarn",
+            "-wlint",
+            "-woff",
+            "819,820,852,821",
+            "-signed",
+            "-DVERSION_JP=1",
+            "-mips2",
+            "-o",
+            temp_o_file,
+            in_file,
+            flag,
+        ]
+    )
+
+
+def do_fix_lohi_step(disasm_info: DisassemblyInfo) -> bytes:
+    logging.debug("Fixing %lo and %hi macros...")
+
+    def replace_last(s: bytes, a: bytes, b: bytes) -> bytes:
+        return b.join(s.rsplit(a, 1))
+
+    output: List[bytes] = []
+    waiting_hi: Dict[bytes, Tuple[int, int, bytes]] = {}
+    for line in disasm_info.disasm.split(b"\n"):
+        line = line.strip()
+        if (disasm_info.entry + b" ")[2:].upper() in line:
+            output.append(b"glabel test")
+
+        index = len(output)
+        if line.startswith(b"func"):
+            line = b"glabel " + line[:-1]
+        elif b"lui" in line:
+            reg, imm_hex = line.split(b" ")[-2::]
+            imm = int(imm_hex, 0)
+            if 0x10 <= imm < 0x1000:  # modestly large lui is probably a pointer
+                waiting_hi[reg[:-1]] = (index, imm, imm_hex)
+        else:
+            lo_regs = [reg for reg in waiting_hi.keys() if reg in line]
+            if lo_regs and "0x" in line:
+                lo_reg = lo_regs[0]
+                hi_ind, hi_imm, hi_imm_hex = waiting_hi[lo_reg]
+                lo_imm_hex = line.split(b" ")[-1].split(b"(")[0]
+                lo_imm = int(lo_imm_hex, 0)
+                if lo_imm >= 0x8000:
+                    lo_imm -= 0x10000
+                sym = b"D_" + bytes(hex((hi_imm << 16) + lo_imm)[2:].upper(), "utf-8")
+                line = replace_last(line, lo_imm_hex, b"%lo(" + sym + b")")
+                output[hi_ind] = replace_last(
+                    output[hi_ind], hi_imm_hex, b"%hi(" + sym + b")"
+                )
+                del waiting_hi[lo_reg]
+        output.append(line)
+    return b"\n".join(output)
+
+
+def irix_compile_with_flag(in_file: Path, out_file: Path, flag: str) -> None:
+    logging.info(f"Compiling {in_file} to {out_file} using this flag: {flag}")
+    with ExitStack() as stack:
+        temp_o_file = stack.enter_context(NamedTemporaryFile(suffix=".o")).name
+        temp_out_file = stack.enter_context(NamedTemporaryFile(suffix=".out")).name
+        logging.debug(f"Compiling and linking {in_file} with {flag}...")
+        do_compilation_step(temp_o_file, str(in_file), flag)
+        do_linker_step(temp_out_file, temp_o_file)
+        disasm_info = do_disassembly_step(temp_out_file)
+        final_asm = do_fix_lohi_step(disasm_info)
+    out_file.write_bytes(final_asm)
+    logging.info(f"Successfully wrote disassembly to {out_file}.")
+
+
+def add_test_from_file(orig_file: Path) -> None:
+    test_dir = Path("end_to_end") / orig_file.stem
+    try:
+        test_dir.mkdir()
+    except FileExistsError:
+        raise Exception(f"{test_dir} already exists. Name your test something else.")

```

Comment:
we need to support test updates as well

Reply:
yeah. Was trying to decide between:
 - Adding an `--update` cli flag
 - Detecting `orig.c` as the filename and/or adding tests that are already in `end_to_end/SOME_DIR/`.